File: src/main/java/io/github/mg138/tsbook/entities/EffectHandler.java
package io.github.mg138.tsbook.entities;

import io.github.mg138.tsbook.entities.data.EntityStatusEffect;
import io.github.mg138.tsbook.entities.data.RegisteredEffects;
import io.github.mg138.tsbook.entities.data.StatusEffect;
import io.github.mg138.tsbook.entities.data.StatusEffectType;
import org.bukkit.entity.Entity;
import org.bukkit.entity.LivingEntity;
import org.bukkit.event.Listener;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.*;

public class EffectHandler implements Listener {
    private final static HashMap<LivingEntity, HashMap<StatusEffectType, BukkitRunnable>> activeRunnable = new HashMap<>();
    private final static HashMap<LivingEntity, HashMap<StatusEffectType, StatusEffect>> activeEffect = new HashMap<>();

    public static void apply(StatusEffectType type, Entity target, double power, int time) {
        RegisteredEffects.Effects.get(type).accept(new EntityStatusEffect(target, type, power, time));
    }

    public static void unload() {
        activeRunnable.clear();
    }

    public static void addEffect(LivingEntity entity, StatusEffect effect) {
        HashMap<StatusEffectType, StatusEffect> effects;

        if (activeEffect.containsKey(entity)) {
            effects = getEffects(entity);
        } else {
            effects = new HashMap<>();
        }

        effects.put(effect.type, effect);
        activeEffect.put(entity, effects);
    }

    public static StatusEffect getEffect(LivingEntity entity, StatusEffectType type) {
        HashMap<StatusEffectType, StatusEffect> effects = getEffects(entity);
        if (effects == null) return null;

        return effects.get(type);
    }

    public static HashMap<StatusEffectType, StatusEffect> getEffects(LivingEntity entity) {
        return activeEffect.get(entity);
    }

    public static boolean hasEffect(LivingEntity entity, StatusEffectType type) {
        HashMap<StatusEffectType, StatusEffect> effects = getEffects(entity);
        if (effects == null) return false;

        return effects.containsKey(type);
    }

    public static void addRunnable(LivingEntity entity, StatusEffect effect, BukkitRunnable runnable) {
        HashMap<StatusEffectType, BukkitRunnable> runnables;

        if (activeRunnable.containsKey(entity)) {
            runnables = activeRunnable.get(entity);
        } else {
            runnables = new HashMap<>();
        }

        BukkitRunnable old = runnables.put(effect.type, runnable);
        if (old != null) old.cancel();

        activeRunnable.put(entity, runnables);
    }

    public static void remove(LivingEntity entity) {
        activeEffect.remove(entity);
        activeRunnable.get(entity).forEach((effect, runnable) -> runnable.cancel());
        activeRunnable.remove(entity);
    }
}

File: src/main/kotlin/io/github/mg138/tsbook/entity/effect/EffectHandler.kt
package io.github.mg138.tsbook.entity.effect

import io.github.mg138.tsbook.Book
import io.github.mg138.tsbook.entity.effect.data.*
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Entity
import java.lang.NullPointerException
import java.util.*
import kotlin.collections.HashMap

object EffectHandler {
    private val activeEffects: MutableMap<LivingEntity, MutableMap<StatusType, EntityEffect>> = HashMap()

    fun getEffect(entity: LivingEntity, type: StatusType) = activeEffects[entity]?.get(type)

    operator fun get(entity: LivingEntity) = activeEffects[entity]

    fun applyEffect(entityEffect: EntityEffect, delay: Long, period: Long) {
        val entityStatus = entityEffect.entityStatus
        val target = entityStatus.target
        if (target !is LivingEntity) return

        activeEffects.putIfAbsent(target, EnumMap<StatusType, EntityEffect>(StatusType::class.java))

        val type = entityStatus.status.type
        this.remove(target, type)
        activeEffects[target]!![type] = entityEffect
        entityEffect.runnable.runTaskTimer(Book.inst, delay, period)
    }

    fun apply(type: StatusType, target: LivingEntity, power: Double, time: Long) {
        RegisteredEffects.effects[type]?.invoke(EntityStatus(target, type, power, time))
    }

    fun removeFromMap(entity: LivingEntity, type: StatusType) {
        activeEffects[entity]?.remove(type)
    }

    fun remove(entity: LivingEntity, type: StatusType) {
        activeEffects[entity]!![type]?.runnable?.cancel()
    }

    fun removeAll(entity: LivingEntity) {
        val removing = try {
            activeEffects[entity]!!
        } catch (e: NullPointerException) {
            throw IllegalArgumentException("The entity doesn't have any active effects")
        }
        removing.forEach { (_, entityEffect) -> entityEffect.runnable.cancel() }
        removing.clear()
        activeEffects.remove(entity)
    }

    fun unload() {
        val removing = HashMap(activeEffects)
        removing.forEach { (_, map) -> map.forEach { (_, entityEffect) -> entityEffect.runnable.cancel() } }
        removing.clear()
        activeEffects.clear()
    }
}